### PR TITLE
Message action sample widget 

### PR DESCRIPTION
**Pull request by qiqingjin**

### CODE REVIEW

File: samples/widgets/message-subscriber/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Message subscriber
+
+This sample demonstrates how to use message action in widget.
```
Comment (dmartine) [inline samples/widgets/message-subscriber/README.md:3]:
```suggestion
This sample demonstrates how to use message action in a widget.
```

File: samples/widgets/message-subscriber/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+## How to use the sample
+Clone the [sample repo](https://github.com/esri/arcgis-experience-builder-sdk-resources) and copy this widget's folder (within `samples/widgets`) to the `client/your-extensions/widgets` folder of your Experience Builder installation.
+
+## How it works
+  Within `query-action-setting.tsx`, create a class extends the React.PureComponent class with the types `ActionSettingProps`.
```
Comment (dmartine) [inline samples/widgets/message-subscriber/README.md:9]:
```suggestion
  Within `query-action-setting.tsx`, create a class to extend the React.PureComponent class with the types `ActionSettingProps`.
```

File: samples/widgets/message-subscriber/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    mainDataSourceId: this.props.config.useDataSource.mainDataSourceId,
+    dataViewId: this.props.config.useDataSource.dataViewId,
+    rootDataSourceId: this.props.config.useDataSource.rootDataSourceId,
+    fields: allSelectedFields.map(f => f.jimuName)
+  })
+});
+
+```
+
+Within `query-action.tsx`, create a class extends the `AbstractMessageAction` class. Then set the URI of the action setting component through the `getSettingComponentUri` method. Use the `filterMessageType` method to filter out actions that are not relevant to this widget. The `onExecute` method is called when the action is triggered.
```
Comment (dmartine) [inline samples/widgets/message-subscriber/README.md:26]:
```suggestion
Within `query-action.tsx`, create a class to extend the `AbstractMessageAction` class. Then set the URI of the action setting component through the `getSettingComponentUri` method. Use the `filterMessageType` method to filter out actions that are not relevant to this widget. The `onExecute` method is called when the action is triggered.
```

File: samples/widgets/message-subscriber/src/runtime/widget.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+  render(){
+    if(!this.isDsConfigured()){
+      return <h3>
+        Please config data source.
+      </h3>;
+    }
+
+    return <div className="widget-subscribe" style={{overflow: 'auto', maxHeight: '700px'}}>
+      <h5>This widget will listen <b>STRING_SELECTION_CHANGE</b> and <b>FEATURE_SELECTION_CHANGE</b> message to do query</h5>
```
Comment (dmartine) [inline samples/widgets/message-subscriber/src/runtime/widget.tsx:43]:
```suggestion
      <h5>This widget will listen to a <b>STRING_SELECTION_CHANGE</b> and <b>FEATURE_SELECTION_CHANGE</b> message to run a query</h5>
```

**Review by dmartine — APPROVED** [review]

File: samples/widgets/message-subscriber/src/actions/query-action-setting.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+interface Config {
+  useDataSource: UseDataSource;
+}
+
+export type IMConfig = ImmutableObject<Config>;
+
+class QueryActionSetting extends React.PureComponent<ActionSettingProps<IMConfig>, States>{
+  constructor(props){
+    super(props);
```
Comment (qlqllu) [inline samples/widgets/message-subscriber/src/actions/query-action-setting.tsx:16]: We should clean this.

File: samples/widgets/message-subscriber/src/actions/query-action-setting.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    this.props.onSettingChange({
+      actionId: this.props.actionId,
+      config: this.props.config.set('useDataSource', initConfig.useDataSource)
+    });
+  }
+
+  /**
+   * Get the selected DataSource from subscribe-widget through widgetId.
+   *
+   * @param widgetId The id of the widget that listens for the message.
```
Comment (qlqllu) [inline samples/widgets/message-subscriber/src/actions/query-action-setting.tsx:66]: should be `listens to`

**Review by qlqllu — APPROVED** [review]

**Pull request merged**